### PR TITLE
GTJ-54

### PR DESCRIPTION
**Pull request by kkeszthe**

### CODE REVIEW

**Comment by kkeszthe** [comment]
@adamgyulavari nézz rá kérlek a jogosultságomra, mert még tudnék mergelni. Vagy az a heroku-s történet miatt van?

File: backend/src/repos/TroopRepo.js [hunk 1]
```diff
@@ -1,36 +1,56 @@
...
-        const query = this.validateQuery`
+  async add({
+    kingdom_id,
+    level,
+    hp,
+    attack,
+    defence,
+    started_at,
+    finished_at,
+  }) {
```
Comment (adamgyulavari) [inline backend/src/repos/TroopRepo.js:29]: @kkeszthe this can be a `Troop` class

File: backend/src/repos/TroopRepo.js [hunk 2]
```diff
@@ -1,36 +1,56 @@
...
+    if (!kingdom_id) throw new Error(this.errorCodes.missingKingdomId);
+    const query = this.validateQuery`
             INSERT INTO kingdom_troops
             (kingdom_id,level,hp,attack,defence,started_at,finished_at)
```
Comment (adamgyulavari) [inline backend/src/repos/TroopRepo.js:33]: @kkeszthe if you introduced `this.table.name` and `this.table.columns`, you should use those in the queries

**Review by adamgyulavari — CHANGES_REQUESTED** [review]
@kkeszthe this format changing is going to kill me :D

**Comment by adamgyulavari** [comment]
> @adamgyulavari nézz rá kérlek a jogosultságomra, mert még tudnék mergelni. Vagy az a heroku-s történet miatt van?

repo admin vagy, adminkent tudsz mergelni by default

File: backend/src/repos/TroopRepo.js (hunk 2 above)
Comment (kkeszthe) [reply backend/src/repos/TroopRepo.js:33]: @adamgyulavari I tried but it breaks the SQL request. The Query handler replaces everything that is marked with ${} a ? and add them to the list of data at the end of the request. SQL was very cross with me about that.

Comment (adamgyulavari) [reply backend/src/repos/TroopRepo.js:33]: I don't see that in the query handler's code

Comment (kkeszthe) [reply backend/src/repos/TroopRepo.js:33]: @adamgyulavari this is the one:
validateQuery(query,...params) {
        params.forEach( param => {
            if (param === undefined) throw new Error(this.errorCodes.missingParam);
        });
        return [
            query.join('?'),
            params
        ];
    };

File: backend/src/services/TroopService.js [hunk 3]
```diff
@@ -1,55 +1,100 @@
...
+      throw new Error(this.errorCodes.invalidTroopAmount);
+
+    this.setTroopData({ kingdomId, level: parseInt(level) + 1, upgrade: true });
+
+    for (let i = 0; i < amount; i++) {
+      if (i === 0) this.troopData.id = [];
+      this.troopData.id.push(upgradebleTroops[i].id);
+    }
+
+    await this.troop.update(this.troopData);
```
Comment (adamgyulavari) [inline backend/src/services/TroopService.js:95]: @kkeszthe 
```suggestion
    await this.troop.update(this.troopData, ids);
```

File: backend/src/services/TroopService.js [hunk 4]
```diff
@@ -1,55 +1,100 @@
...
+    ).troops.filter(troop => troop.level == level);
+    if (upgradebleTroops.length < amount)
+      throw new Error(this.errorCodes.invalidTroopAmount);
+
+    this.setTroopData({ kingdomId, level: parseInt(level) + 1, upgrade: true });
+
+    for (let i = 0; i < amount; i++) {
+      if (i === 0) this.troopData.id = [];
+      this.troopData.id.push(upgradebleTroops[i].id);
+    }
```
Comment (adamgyulavari) [inline backend/src/services/TroopService.js:93]:
```suggestion
    const ids = upgradebleTroops.map(t => t.id);
```

**Review by adamgyulavari — DISMISSED** [review]

**Comment by kkeszthe** [comment]
@adamgyulavari pretty please approve it, so I can merge it in. :-)

**Review by adamgyulavari — APPROVED** [review]

**Pull request merged**